Roster acceptance test.

HassanJaveed84 opened the pull request:
@mulby @brianhw 

Up for initial review, I think we should add more data to the input event files. 
Right now, only one test user has an enrollment event.

mulby commented on config/test.cfg:
```diff
@@ -114,3 +114,11 @@ cybersource_merchant_ids = test
 dropoff_threshold = 0.05
+
+[elasticsearch]
+host = https://search-acceptance-test-kkpt4pnwnib3ywbu3ovsselg6y.us-east-1.es.amazonaws.com/
```
We should pass this into the acceptance test framework in the ACCEPTANCE_TEST_CONFIG instead of committing it to the repo.

mulby replied:
https://github.com/edx/edx-analytics-pipeline/blob/master/edx/analytics/tasks/tests/acceptance/__init__.py#L149

mulby commented on edx/analytics/tasks/tests/acceptance/test_module_engagement.py:
```diff
@@ -19,6 +20,53 @@ class ModuleEngagementAcceptanceTest(AcceptanceTestCase):
+        self.execute_sql_fixture_file('load_course_groups_courseusergroup_users.sql')
+
+        self.task.launch([
+            'ModuleEngagementRosterIndexTask',
```
We should use this entry point:

https://github.com/edx/edx-analytics-pipeline/blob/master/edx/analytics/tasks/module_engagement.py#L1189

And also verify the metric ranges since both are required for Insights to work properly.

mulby commented:
Directionally, this is perfect. I've suggested a couple of tweaks.

I also think having a couple of users would be very useful. This is the only coverage we have of the hive query, so I think it's worth adding a little test complexity to give us more assurance that query continues to do what we want it to.

HassanJaveed84 commented:
@mulby @brianhw can you please review this while I add in more data.

mulby commented:
@HassanJaveed84 I think the double-run issue may be related to --overwrite. I would remove it from your test.

mulby commented on edx/analytics/tasks/module_engagement.py:
```diff
@@ -1187,6 +1187,7 @@ def document_generator(self, lines):
 
 @workflow_entry_point  # pylint: disable=missing-docstring
 class ModuleEngagementWorkflowTask(ModuleEngagementDownstreamMixin,
+                                   OverwriteOutputMixin,
```
This mixin messes with the `complete()` method for the task. We should just manually define an overwrite parameter here...

HassanJaveed84 replied:
ModuleEngagementSummaryMetricRangesMysqlTask defines an overwrite with a default value of True. RosterIndexTask has a default of False for overwrite. 
Do we need to define two different overwrite parameters for the workflow ?

mulby replied:
I think it's OK to not pass overwrite to the metric ranges, since we pretty much always want that to overwrite.

We need to do some architectural thinking about overwrite... the current approach is very error prone and non-intuitive.

mulby commented:
@HassanJaveed84 I think I fixed the run-twice issue in this PR: https://github.com/edx/edx-analytics-pipeline/pull/238

It looks like we both fixed the overwrite parameter... I'll strip the change out of my branch.

mulby commented on edx/analytics/tasks/tests/acceptance/test_module_engagement.py:
```diff
@@ -19,7 +22,119 @@ class ModuleEngagementAcceptanceTest(AcceptanceTestCase):
+            'ModuleEngagementWorkflowTask',
+            '--date', '2015-04-17',
+            '--n-reduce-tasks', str(self.NUM_REDUCERS),
+            '--overwrite'
```
I would remove "--overwrite" here since that's not really the standard way this will be invoked.

mulby commented on edx/analytics/tasks/tests/acceptance/test_module_engagement.py:
```diff
@@ -19,7 +22,119 @@ class ModuleEngagementAcceptanceTest(AcceptanceTestCase):
     NUM_REDUCERS = 1
 
-    def test_engagement(self):
+    def test_roster_generation(self):
```
We should add a decorator around this that prevents it from being run locally, much like the decorators for vertica, s3 etc.

mulby commented on edx/analytics/tasks/tests/acceptance/test_module_engagement.py:
```diff
@@ -19,7 +22,119 @@ class ModuleEngagementAcceptanceTest(AcceptanceTestCase):
+            ('edX/DemoX/Demo_Course_2', datetime.date(2015, 4, 10), datetime.date(2015, 4, 17), 'videos_viewed', 'low', 0, 1.0),
+            ('edX/DemoX/Demo_Course_2', datetime.date(2015, 4, 10), datetime.date(2015, 4, 17), 'videos_viewed', 'normal', 1.0, 1.0),
+            ('edX/DemoX/Demo_Course_2', datetime.date(2015, 4, 10), datetime.date(2015, 4, 17), 'videos_viewed', 'high', 1.0, None),
+        ])
```
we should also verify the contents of the `module_engagement` table.

mulby replied:
We might be able to pull a bunch of that logic from the `test_engagement` test below. Once we do so, we can remove that test.

mulby commented:
@HassanJaveed84 - I made a pass over this test. Generally looks good. I had a few comments that I think we should address before merging.

mulby commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -169,7 +169,10 @@ def setUp(self):
-            }
+            },
+            'elasticsearch': {
+                'host': self.config['elasticsearch_host']
```
this should be an optional config setting... if it's not specified than this section shouldn't be included (kind of like 'vertica_creds_url')

HassanJaveed84 commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -247,3 +272,17 @@ def execute_sql_fixture_file(self, sql_file_name, database=None):
+                    self.elasticsearch.indices.delete(index=index)
+
+        #get documents from the marker index that have their target_index set to current alias
+        query = {"query": {"wildcard": {"target_index": '*_{}'.format(self.elasticsearch_alias)}}}
```
We can change our test alias to be of the form  'random_string' + '_roster' and then use a prefix query here instead of wildcard.

HassanJaveed84 commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -247,3 +272,17 @@ def execute_sql_fixture_file(self, sql_file_name, database=None):
+
+        #get documents from the marker index that have their target_index set to current alias
+        query = {"query": {"wildcard": {"target_index": '*_{}'.format(self.elasticsearch_alias)}}}
+        response = self.elasticsearch_client.search(index='index_updates', body=query)
```
This won't return all results, but there would be only one document anyway.
We can also use 'scroll' instead.

HassanJaveed84 commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -247,3 +272,17 @@ def execute_sql_fixture_file(self, sql_file_name, database=None):
+        response = self.elasticsearch_client.search(index='index_updates', body=query)
+
+        for doc in response['hits']['hits']:
+            self.elasticsearch_client.delete(index='index_updates', doc_type='marker', id=doc['_id'])
```
We can change this to use bulk delete api.

mulby commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -208,6 +229,9 @@ def setUp(self):
         self.vertica = vertica.VerticaService(self.config, schema)
 
+        # This supports only aws es clusters for now.
+        self.elasticsearch_client = elasticsearch.Elasticsearch(hosts=[self.config['elasticsearch_host']], connection_class=AwsHttpConnection)
```
I think it would make sense to move most of this logic to a new "ElasticsearchService" similar to the HiveService and VerticaService etc.

It could contain all of the logic around deleting the indexes and index_updates document etc, as well as making it easier for the test to run queries etc.

mulby commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -7,8 +7,14 @@
+
 from edx.analytics.tasks.tests.acceptance.services import fs, db, task, hive, vertica
 from edx.analytics.tasks.url import url_path_join, get_target_from_url
+from edx.analytics.tasks.util.aws_elasticsearch_connection import AwsHttpConnection
```
Ideally this would be configurable, eventually we would like to support running this acceptance tests on devstack which would want to use the default connection class.

mulby commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -247,3 +272,17 @@ def execute_sql_fixture_file(self, sql_file_name, database=None):
         log.debug('Executing SQL fixture %s on %s', sql_file_name, database.database_name)
         database.execute_sql_file(os.path.join(self.data_dir, 'input', sql_file_name))
+
+    def reset_es(self):
```
I'm actually OK with this method as is. Just put in comments that call out the fact that there should only be 1 document most of the time.

These reset() methods don't need to be terribly performant so I wouldn't worry about optimizing them very much.

HassanJaveed84 commented on edx/analytics/tasks/tests/acceptance/services/es.py:
```diff
@@ -0,0 +1,31 @@
+class ElasticsearchService(object):
+
+    def __init__(self, config, alias_suffix):
+         self.elasticsearch_client = elasticsearch.Elasticsearch(hosts=[config['elasticsearch_host']], connection_class=eval(config.get('elasticsearch_connection_class', 'AwsHttpConnection')))
```
Is there a way to substitute eval for something else ?
If the connection_class is configurable, how do we dynamically import it ?

mulby replied:
I just had a hardcode mapping:

```python
if config.get('elasticsearch_connection_class') == 'aws':
    connection_class = AwsHttpConnection
```

Not pretty, but probably fine for test code.

mulby replied:
You can probably do:

```python
if config.get('elasticsearch_connection_class') == 'aws':
    connection_class = AwsHttpConnection
else:
    connection_class = None
```

HassanJaveed84 commented on edx/analytics/tasks/tests/acceptance/services/es.py:
```diff
@@ -0,0 +1,31 @@
+         self.alias_suffix = alias_suffix
+
+    @property
+    def client(self):
```
Should there be some sort of a search method in this service that invokes ES API instead of using a property in tests ?

mulby replied:
I think the way you have it is fine!

mulby commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -207,6 +222,7 @@ def setUp(self):
         self.task = task.TaskService(self.config, task_config_override, self.identifier)
         self.hive = hive.HiveService(self.task, self.config, database_name)
         self.vertica = vertica.VerticaService(self.config, schema)
+        self.es = es.ElasticsearchService(self.config, elasticsearch_alias_suffix)
```
nit, I would prefer you pass in the entire alias name here instead of just the suffix

mulby commented on edx/analytics/tasks/tests/acceptance/test_module_engagement.py:
```diff
@@ -19,56 +22,180 @@ class ModuleEngagementAcceptanceTest(AcceptanceTestCase):
-            )
-            results = cursor.fetchall()
+        #es_host = self.task.default_config_override['elasticsearch']['host']
+        #es_client = elasticsearch.Elasticsearch(hosts=[es_host], connection_class=AwsHttpConnection)
```
OK to remove these, I like the new way better.

mulby commented on edx/analytics/tasks/tests/acceptance/test_module_engagement.py:
```diff
@@ -19,56 +22,180 @@ class ModuleEngagementAcceptanceTest(AcceptanceTestCase):
-            (april_thirteenth, 'edX/DemoX/Demo_Course', 'honor', 'video', 'i4x-edX-DemoX-video-8c0028eb2a724f48a074bc184cd8635f', 'viewed', 2),
-            (april_thirteenth, 'edX/DemoX/Demo_Course_2', 'honor', 'video', 'i4x-edX-DemoX-video-8c0028eb2a724f48a074bc184cd8635f', 'viewed', 1),
-            (april_thirteenth, 'edX/DemoX/Demo_Course_2', 'staff', 'video', 'i4x-edX-DemoX-video-8c0028eb2a724f48a074bc184cd8635f', 'viewed', 1),
+        roster_alias = self.task.default_config_override['module-engagement']['alias']
```
I would prefer we move this logic to the ElasticsearchService

brianhw commented on edx/analytics/tasks/elasticsearch_load.py:
```diff
@@ -150,6 +151,8 @@ def init_local(self):
                 )
             )
 
+        self.attempted_removal = True
```
This may be obscure enough that it would benefit from a comment.

mulby replied:
I just pushed a commit to add a comment here.

@HassanJaveed84 you can pull my commit when you pick this up again. I figured since this was part of my original patch, I should fix it.

brianhw commented on edx/analytics/tasks/module_engagement.py:
```diff
@@ -1249,15 +1248,18 @@ class ModuleEngagementWorkflowTask(ModuleEngagementDownstreamMixin,
         description=ModuleEngagementRosterIndexTask.obfuscate.description
     )
 
+    overwrite = luigi.BooleanParameter(default=False)
```
This also needs a comment, so I have a clue as to why it's here.  (As opposed to including OverwriteOutputMixin.)

mulby replied:
Done.

brianhw commented on edx/analytics/tasks/tests/acceptance/services/es.py:
```diff
@@ -0,0 +1,31 @@
+                if alias.endswith(self.alias_suffix):
+                    self.elasticsearch.indices.delete(index=index)
+
+        #get documents from the marker index that have their target_index set to current alias
```
nit: space after #, init cap, end with period.

brianhw commented on edx/analytics/tasks/tests/acceptance/test_module_engagement.py:
```diff
@@ -19,56 +22,180 @@ class ModuleEngagementAcceptanceTest(AcceptanceTestCase):
+        query = {"query": {"match_all": {}}}
+        response = self.es.client.search(index=roster_alias, body=query)
+
+        self.maxDiff = None
```
If this is just for debugging, consider removing it....

HassanJaveed84 replied:
Done.

brianhw commented on edx/analytics/tasks/tests/acceptance/test_module_engagement.py:
```diff
@@ -19,56 +22,180 @@ class ModuleEngagementAcceptanceTest(AcceptanceTestCase):
+
+        self.assertItemsEqual([
+            row for row in results
+        ], [
```
nit:  I would put the list comprehension on a  single line:  
self.assertItemsEqual(
     [row for row in results],
    [
...
    ]
)

HassanJaveed84 replied:
Done

brianhw commented:
Finished a pass -- only nits found.

HassanJaveed84 commented:
Updated some stuff:

- The acceptance test generates the full alias and passes to ElasticsearchService.
- Simplified ElasticsearchService's reset as it has full alias(uses match instead of wildcard)
- Made an 'alias' property in ElasticsearchService(the acceptance test uses it).

mulby commented on edx/analytics/tasks/tests/acceptance/services/es.py:
```diff
@@ -0,0 +1,42 @@
+try:
```
nit: I think it would be clearer if we called the service "elasticsearch" instead of "es"

mulby commented:
One nit. This is ready to merge as far as I'm concerned.

:+1:

brianhw commented:
LGTM.  :+1:

HassanJaveed84 commented:
Had some merge conflicts during rebase(in module_engagement.py). 
Just want to config that I resolved them correctly before merging.

mulby commented:
:+1: merge looks good to me
